### PR TITLE
Minor Bug Fixes

### PR DESCRIPTION
**Pull request by shehzan10**
* Fix syncthreads missed from #1595 
* Fix double free issue in af_assign gen

[skip arrayfire ci]

### CODE REVIEW

File: src/backend/cuda/kernel/nearest_neighbour.hpp [hunk 1]
```diff
@@ -411,9 +411,9 @@ __global__ void select_matches(
                 s_dist[sid] = dist;
                 s_idx[sid]  = s_idx[sid + i];
             }
-            __syncthreads();
         }
```
Comment (pavanky) [inline src/backend/cuda/kernel/nearest_neighbour.hpp]: `__syncthreads()` should go here. This is a block reduction step.

Comment (shehzan10) [inline src/backend/cuda/kernel/nearest_neighbour.hpp]: Ah! my bad. i thought i moved it there.

Comment (shehzan10) [inline src/backend/cuda/kernel/nearest_neighbour.hpp]: Corrected.

**Comment by pavanky** [comment]
@shehzan10 can you also remove this printf
https://github.com/arrayfire/arrayfire/blob/hotfix-3.4.1/src/backend/opencl/kernel/bilateral.hpp#L53

**Comment by pavanky** [comment]
Also this. 

```
diff --git a/src/backend/cpu/scan_by_key.cpp b/src/backend/cpu/scan_by_key.cpp
index d00db6c..f176840 100644
--- a/src/backend/cpu/scan_by_key.cpp
+++ b/src/backend/cpu/scan_by_key.cpp
@@ -31,6 +31,7 @@ namespace cpu
         kernel::scan_dim_by_key<op, Ti, Tk, To, 4> func4(inclusive_scan);
 
         in.eval();
+        key.eval();
         switch (in.ndims()) {
         case 1:
             getQueue().enqueue(func1, out, 0, key, 0, in, 0, dim);
```

**Pull request merged**